posthog/models/person/util.py

```diff
@@ -148,25 +149,39 @@ def get_persons_by_uuids(team: Team, uuids: List[str]) -> QuerySet:
     return Person.objects.filter(team_id=team.pk, uuid__in=uuids)
 
 
-def delete_person(
-    person_id: UUID, properties: Dict, is_identified: bool, delete_events: bool = False, team_id: int = False
-) -> None:
+def delete_person(person: Person, delete_events: bool = False, delete_distinct_ids=False) -> None:
     timestamp = now()
 
     data = {
-        "id": person_id,
-        "team_id": team_id,
-        "properties": json.dumps(properties),
-        "is_identified": int(is_identified),
+        "id": person.uuid,
+        "team_id": person.team.id,
+        "properties": "{}",
+        "is_identified": int(person.is_identified),
         "created_at": timestamp.strftime("%Y-%m-%d %H:%M:%S"),
         "_timestamp": timestamp.strftime("%Y-%m-%d %H:%M:%S"),
+        "version": int(person.version or 0) + 100,  # keep in sync with deletePerson in plugin-server/src/utils/db/db.ts
     }
 
-    try:
-        if delete_events:
-            sync_execute(DELETE_PERSON_EVENTS_BY_ID, {"id": person_id, "team_id": team_id})
-    except:
-        pass  # cannot delete if the table is distributed
+    if delete_events:
+        sync_execute(
+            DELETE_PERSON_EVENTS_BY_ID.format(events_data_table=EVENTS_DATA_TABLE(), cluster=CLICKHOUSE_CLUSTER),
+            {"id": person.uuid, "team_id": person.team.id},
+            settings={"mutations_sync": 1 if TEST else 0},
+        )
+
+    if delete_distinct_ids:
+        distinct_id_inserts = []
+        distinct_id_map: Dict[str, str] = {}
+        for i, distinct_id in enumerate(person.distinct_ids):
```

Comment:
Isn't this prone to race conditions? An ALTER TABLE DELETE with subquery _seems_ better here.

Reply:
It is - tradeoff is yet another mutation

Reply:
True :thinking: that's not ideal. Another option is to use nested selects in a INSERT INTO as well.

Reply:
Actually thinking about it now this is fine - we might send a message mapping a distinct_id to a new person_id, but won't send a new message for the same distinct_id:person_id (+ team_id) mapping. Thus this should work. Also if some IDs are no longer associated to this person by the time we do this, there's also no harm done